### PR TITLE
Correct the source location of methods and not null annotation checks [TG-8270]

### PR DESCRIPTION
**Pull request by majakusber**
https://github.com/diffblue/cbmc/pull/4802 introduced not-null-annotation checks for input parameters. These properties are added at the beginning of the goto program of the method. The source location of the property was missing information - line  number and method name. The information is inherited from the method itself, so as a fix I set the line and source location of the method itself.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8df4a6b20f7a9c744eef657e5cea63c739af46ed).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118811498

**Review by smowton — APPROVED** [review]

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 1]
```diff
@@ -1256,10 +1256,23 @@ void java_bytecode_parsert::rmethod_attribute(methodt &method)
...
+        method.instructions.end(),
+        [&](const instructiont &instruction) {
+          return !instruction.source_location.get_line().empty();
+        });
+      if(it != method.instructions.end())
+        method.source_location.set_line(it->source_location.get_line());
+    }
+
+    // method name
+    method.source_location.set_function(
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Might be worth factoring out into a separate function as this code is duplicated here: https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/java_bytecode_parser.cpp#L1254

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 2]
```diff
@@ -1256,10 +1256,23 @@ void java_bytecode_parsert::rmethod_attribute(methodt &method)
           id2string(method.name)+":"+method.descriptor);
     }
 
-    // line number of method
+    // line number of method (the first line number available)
     if(!method.instructions.empty())
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Isn't that covered by the check in line 1268?

**Review by peterschrammel — APPROVED** [review]

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp (hunk 2 above)
Comment (majakusber) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Indeed, fixed

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp (hunk 1 above)
Comment (majakusber) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Done, I moved the assignment to he method location higher up, and then just re-use it's value when assigning the instruction location.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 652484972222c9f5a39b948e7ed80392969a38e2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118920380
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 85a1e8b3bcea7e78fd5515a83a847049f4f9753a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118920462

**Pull request merged**